File: app-macos/Sources/RunnerForge/Services/GitHubAppService.swift
import Foundation
import Security

/// A single-use runner registration. The blob never touches disk and never
/// appears in an argument list.
public struct JitConfig: Sendable {
    public let runnerName: String
    public let encodedConfig: String
}

public enum GitHubError: Error, CustomStringConvertible {
    case noPrivateKey
    case badPrivateKey(String)
    case api(status: Int, body: String, operation: String)
    case malformedResponse(String)

    public var description: String {
        switch self {
        case .noPrivateKey:
            "No GitHub App private key is stored. Add it on the Credentials page; it is kept in the "
            + "Keychain, never in forge.json."
        case .badPrivateKey(let detail):
            "The stored GitHub App key could not be used: \(detail)"
        case .api(let status, let body, let operation):
            // GitHub's own message, verbatim. Guessing what a 401 means wastes
            // far more time than reading what GitHub actually said.
            "Could not \(operation) (\(status)). GitHub said:\n\(body)"
        case .malformedResponse(let detail):
            "Unexpected response from GitHub: \(detail)"
        }
    }
}

/// Turns a GitHub App private key into short-lived credentials, and mints the
/// single-use JIT config every runner registers with.
///
/// There is no long-lived registration token anywhere in this product, and no
/// `config.sh` step: a JIT config IS the configuration, valid for one job.
public actor GitHubAppService {
    private let logBus: LogBus
    private let secretStore: SecretStore
    private let session: URLSession

    private var cachedToken: String?
    private var cachedTokenExpiry: Date = .distantPast

    public var apiBaseUrl = "https://api.github.com"

    public init(logBus: LogBus, secretStore: SecretStore, session: URLSession = .shared) {
        self.logBus = logBus
        self.secretStore = secretStore
        self.session = session
    }

    /// Signs an app JWT. `iat` is backdated 60 seconds to absorb clock skew,
    /// which GitHub otherwise rejects as a token issued in the future, and `exp`
    /// is 9 minutes, inside GitHub's 10-minute maximum.
    public nonisolated func createAppJwt(appId: String, privateKeyPem: String) throws -> String {
        guard privateKeyPem.contains("PRIVATE KEY") else {
            throw GitHubError.badPrivateKey("it is not a PEM private key")
        }

        let now = Int(Date().timeIntervalSince1970)
        let header = #"{"alg":"RS256","typ":"JWT"}"#
        let payload = #"{"iat":\#(now - 60),"exp":\#(now + 540),"iss":"\#(appId)"}"#

        let signingInput = Self.base64Url(Data(header.utf8)) + "." + Self.base64Url(Data(payload.utf8))

        // RSA signing goes through the Security framework. CryptoKit exposes no
        // RSA on Apple platforms, so SecKey is the supported path — and it is
        // the same primitive the OS uses, rather than a bundled reimplementation.
        let key = try Self.secKey(fromPem: privateKeyPem)

        var error: Unmanaged<CFError>?
        guard let signature = SecKeyCreateSignature(
            key,
            .rsaSignatureMessagePKCS1v15SHA256,
            Data(signingInput.utf8) as CFData,
            &error) as Data? else {
            let detail = (error?.takeRetainedValue()).map { String(describing: $0) } ?? "unknown error"
            throw GitHubError.badPrivateKey("signing failed: \(detail)")
        }

        return signingInput + "." + Self.base64Url(signature)
    }

    /// Parses a PEM private key into a SecKey.
    ///
    /// GitHub App keys are PKCS#1 ("BEGIN RSA PRIVATE KEY"). A PKCS#8 key
    /// ("BEGIN PRIVATE KEY") wraps the same PKCS#1 structure behind a fixed
    /// 26-byte header for RSA, which is stripped here so both forms work —
    /// people do convert their keys, and failing on one form is a confusing bug.
    nonisolated static func secKey(fromPem pem: String) throws -> SecKey {
        let base64 = pem
            .split(separator: "\n")
            .filter { !$0.hasPrefix("-----") }
            .joined()

        guard var der = Data(base64Encoded: base64, options: [.ignoreUnknownCharacters]) else {
            throw GitHubError.badPrivateKey("the PEM body is not valid base64")
        }

        if pem.contains("BEGIN PRIVATE KEY") {
            // PKCS#8 RSA header: SEQUENCE, version, rsaEncryption OID, NULL,
            // OCTET STRING. 26 bytes before the embedded PKCS#1 key.
            let pkcs8HeaderLength = 26
            guard der.count > pkcs8HeaderLength else {
                throw GitHubError.badPrivateKey("the PKCS#8 key is too short to contain a key")
            }
            der = der.dropFirst(pkcs8HeaderLength)
        }

        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass as String: kSecAttrKeyClassPrivate,
        ]

        var error: Unmanaged<CFError>?
        guard let key = SecKeyCreateWithData(der as CFData, attributes as CFDictionary, &error) else {
            let detail = (error?.takeRetainedValue()).map { String(describing: $0) } ?? "unknown error"
            throw GitHubError.badPrivateKey("could not parse the key: \(detail)")
        }

        return key
    }

    nonisolated static func base64Url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    /// An installation access token, cached for 50 minutes. GitHub issues them
    /// for 60; the margin means a token never expires mid-use.
    public func installationToken(config: ForgeConfig) async throws -> String {
        if let cachedToken, Date() < cachedTokenExpiry { return cachedToken }

        guard let privateKey = await secretStore.read("githubAppPrivateKey") else {
            throw GitHubError.noPrivateKey
        }

        let jwt = try createAppJwt(appId: config.github.appId, privateKeyPem: privateKey)

        let (data, response) = try await send(
            path: "/app/installations/\(config.github.installationId)/access_tokens",
            method: "POST", token: jwt, body: nil)

        guard (200..<300).contains(response.statusCode) else {
            throw GitHubError.api(status: response.statusCode,
                                  body: String(decoding: data, as: UTF8.self),
                                  operation: "mint an installation token")
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = json["token"] as? String else {
            throw GitHubError.malformedResponse("the token response contained no token")
        }

        await logBus.registerSecret(token)
        cachedToken = token
        cachedTokenExpiry = Date().addingTimeInterval(50 * 60)
        await logBus.info("github", "minted an installation token (cached 50 minutes)")

        return token
    }

    /// Mints a JIT config for one runner. The returned blob is the runner's whole
    /// configuration and is valid for exactly one job.
    public func generateJitConfig(
        config: ForgeConfig, repo: String, runnerClass: RunnerClass
    ) async throws -> JitConfig {
        let token = try await installationToken(config: config)

        let shortHost = String(
            ProcessInfo.processInfo.hostName
                .filter { $0.isLetter || $0.isNumber || $0 == "-" }
                .prefix(16)
        ).lowercased()

        let suffix = (0..<4).map { _ in String(format: "%02x", UInt8.random(in: 0...255)) }.joined()
        let runnerName = "forge-\(runnerClass.classId)-\(shortHost)-\(suffix)"

        let configured = config.runners.first { $0.classId == runnerClass.classId }
        let labels = (configured?.labels.isEmpty == false) ? configured!.labels : runnerClass.labels

        let payload: [String: Any] = [
            "name": runnerName,
            "runner_group_id": 1,
            "labels": labels,
        ]

        let (data, response) = try await send(
            path: "/repos/\(config.github.owner)/\(repo)/actions/runners/generate-jitconfig",
            method: "POST", token: token,
            body: try JSONSerialization.data(withJSONObject: payload))

        guard (200..<300).contains(response.statusCode) else {
            throw GitHubError.api(status: response.statusCode,
                                  body: String(decoding: data, as: UTF8.self),
                                  operation: "generate a JIT config")
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let blob = json["encoded_jit_config"] as? String else {
            throw GitHubError.malformedResponse("the response contained no encoded_jit_config")
        }

        await logBus.registerSecret(blob)
        // The LENGTH only. The value must never reach a log.
        await logBus.info("github", "minted a JIT config for \(runnerName) (\(blob.count) bytes)")

        return JitConfig(runnerName: runnerName, encodedConfig: blob)
    }

    /// Verifies the App key by asking GitHub who we are.
    public func testAppCredentials(config: ForgeConfig) async throws -> String {
        guard let privateKey = await secretStore.read("githubAppPrivateKey") else {
            throw GitHubError.noPrivateKey
        }

        let jwt = try createAppJwt(appId: config.github.appId, privateKeyPem: privateKey)
        let (data, response) = try await send(path: "/app", method: "GET", token: jwt, body: nil)

        guard (200..<300).contains(response.statusCode) else {
            throw GitHubError.api(status: response.statusCode,
                                  body: String(decoding: data, as: UTF8.self),
                                  operation: "GET /app")
        }

        let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        return (json?["slug"] as? String) ?? "(the app has no slug)"
    }

    /// Checks repository access, reporting 200/404/403 distinctly.
    public func verifyRepositoryAccess(config: ForgeConfig, repo: String) async throws -> String {
        let token = try await installationToken(config: config)
        let (_, response) = try await send(
            path: "/repos/\(config.github.owner)/\(repo)", method: "GET", token: token, body: nil)

        switch response.statusCode {
        case 200: return "OK — the installation can see this repository."
        case 404: return "404 — either the repository does not exist, or the App is not installed on it. "
                       + "GitHub returns 404 rather than 403 for repositories you cannot see at all."
        case 403: return "403 — the App is installed but lacks permission. It needs Administration: "
                       + "read & write to register self-hosted runners."
        default: return "\(response.statusCode)"
        }
    }

    /// Retries 5xx and secondary-rate-limit 403s with exponential backoff.
    func send(path: String, method: String, token: String, body: Data?) async throws -> (Data, HTTPURLResponse) {
        let maxAttempts = 4

        for attempt in 1...maxAttempts {
            var request = URLRequest(url: URL(string: apiBaseUrl + path)!)
            request.httpMethod = method
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
            request.setValue("2022-11-28", forHTTPHeaderField: "X-GitHub-Api-Version")
            request.setValue("RunnerForge/1.0", forHTTPHeaderField: "User-Agent")
            if let body {
                request.httpBody = body
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            }

            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                throw GitHubError.malformedResponse("not an HTTP response")
            }

            let retryAfter = http.value(forHTTPHeaderField: "Retry-After").flatMap(Double.init)
            let retryable = http.statusCode >= 500 || (http.statusCode == 403 && retryAfter != nil)

            if !retryable || attempt == maxAttempts { return (data, http) }

            let delay = retryAfter ?? pow(2.0, Double(attempt))
            await logBus.warning("github",
                "\(http.statusCode) from \(path); retrying in \(Int(delay))s (attempt \(attempt)/\(maxAttempts))")
            try? await Task.sleep(for: .seconds(delay))
        }

        throw GitHubError.malformedResponse("retry loop exhausted")
    }
}

File: app-windows/RunnerForge/Services/GitHubAppService.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using RunnerForge.Models;

namespace RunnerForge.Services;

/// <summary>
/// Turns a GitHub App private key into short-lived credentials, and mints the
/// single-use JIT config every runner registers with.
/// </summary>
/// <remarks>
/// There is no long-lived registration token anywhere in this product, and no
/// <c>config.cmd</c> step: a JIT config IS the configuration, valid for exactly
/// one job.
/// </remarks>
public sealed class GitHubAppService(LogBus logBus, SecretStore secretStore, HttpClient httpClient)
{
    private readonly LogBus _logBus = logBus;
    private readonly SecretStore _secretStore = secretStore;
    private readonly HttpClient _httpClient = httpClient;

    private string? _cachedToken;
    private DateTimeOffset _cachedTokenExpiry = DateTimeOffset.MinValue;

    public string ApiBaseUrl { get; set; } = "https://api.github.com";

    /// <summary>
    /// Signs an app JWT. <c>iat</c> is backdated 60 seconds to absorb clock skew,
    /// which GitHub otherwise rejects as a token issued in the future, and
    /// <c>exp</c> is 9 minutes, inside GitHub's 10-minute maximum.
    /// </summary>
    public string CreateAppJwt(string appId, string privateKeyPem)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(appId);
        ArgumentException.ThrowIfNullOrWhiteSpace(privateKeyPem);

        if (!privateKeyPem.Contains("PRIVATE KEY", StringComparison.Ordinal))
        {
            throw new InvalidOperationException(
                "The stored GitHub App key is not a PEM private key. Re-enter it on the Credentials page.");
        }

        long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        string header = """{"alg":"RS256","typ":"JWT"}""";
        string payload = $$"""{"iat":{{now - 60}},"exp":{{now + 540}},"iss":"{{appId}}"}""";

        string signingInput = Base64Url(Encoding.UTF8.GetBytes(header))
                              + "." + Base64Url(Encoding.UTF8.GetBytes(payload));

        using var rsa = RSA.Create();
        rsa.ImportFromPem(privateKeyPem);

        byte[] signature = rsa.SignData(
            Encoding.UTF8.GetBytes(signingInput), HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);

        return signingInput + "." + Base64Url(signature);
    }

    private static string Base64Url(byte[] bytes) =>
        Convert.ToBase64String(bytes).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    /// <summary>
    /// An installation access token, cached for 50 minutes. GitHub issues them
    /// for 60; the margin means a token never expires mid-use.
    /// </summary>
    public async Task<string> GetInstallationTokenAsync(
        ForgeConfig config, CancellationToken cancellationToken = default)
    {
        if (_cachedToken is not null && DateTimeOffset.UtcNow < _cachedTokenExpiry)
        {
            return _cachedToken;
        }

        string privateKey = _secretStore.Read("githubAppPrivateKey")
            ?? throw new InvalidOperationException(
                "No GitHub App private key is stored. Add it on the Credentials page; "
                + "it is kept in Windows Credential Manager, never in forge.json.");

        string jwt = CreateAppJwt(config.GitHub.AppId, privateKey);

        using var request = new HttpRequestMessage(
            HttpMethod.Post,
            $"{ApiBaseUrl}/app/installations/{config.GitHub.InstallationId}/access_tokens");
        ApplyHeaders(request, "Bearer", jwt);

        using HttpResponseMessage response =
            await SendWithRetryAsync(request, cancellationToken).ConfigureAwait(false);
        string body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);

        if (!response.IsSuccessStatusCode)
        {
            // GitHub's own message, verbatim. Guessing what a 401 means wastes
            // far more time than reading what GitHub actually said.
            throw new InvalidOperationException(
                $"Could not mint an installation token ({(int)response.StatusCode}). GitHub said:{Environment.NewLine}{body}");
        }

        JsonNode node = JsonNode.Parse(body)
            ?? throw new InvalidOperationException("The token response was not JSON.");
        string token = node["token"]?.GetValue<string>()
            ?? throw new InvalidOperationException("The token response contained no token.");

        _logBus.RegisterSecret(token);
        _cachedToken = token;
        _cachedTokenExpiry = DateTimeOffset.UtcNow.AddMinutes(50);
        _logBus.Info("github", "minted an installation token (cached 50 minutes)");

        return token;
    }

    /// <summary>
    /// Mints a JIT config for one runner. The returned blob is the runner's whole
    /// configuration and is valid for exactly one job.
    /// </summary>
    public async Task<JitConfig> GenerateJitConfigAsync(
        ForgeConfig config,
        string repo,
        RunnerClass runnerClass,
        CancellationToken cancellationToken = default)
    {
        string token = await GetInstallationTokenAsync(config, cancellationToken).ConfigureAwait(false);

        string shortHost = new string(Environment.MachineName
            .Where(char.IsLetterOrDigit).Take(16).ToArray()).ToLowerInvariant();
        string suffix = Convert.ToHexString(RandomNumberGenerator.GetBytes(4)).ToLowerInvariant();
        string runnerName = $"forge-{runnerClass.ClassId}-{shortHost}-{suffix}";

        RunnerConfig? configured = config.Runners.FirstOrDefault(r => r.ClassId == runnerClass.ClassId);
        IReadOnlyList<string> labels = configured?.Labels is { Count: > 0 }
            ? configured.Labels : runnerClass.Labels;

        var payload = new JsonObject
        {
            ["name"] = runnerName,
            ["runner_group_id"] = 1,
            ["labels"] = new JsonArray([.. labels.Select(l => (JsonNode)JsonValue.Create(l)!)]),
        };

        using var request = new HttpRequestMessage(
            HttpMethod.Post,
            $"{ApiBaseUrl}/repos/{config.GitHub.Owner}/{repo}/actions/runners/generate-jitconfig")
        {
            Content = new StringContent(payload.ToJsonString(), Encoding.UTF8, "application/json"),
        };
        ApplyHeaders(request, "Bearer", token);

        using HttpResponseMessage response =
            await SendWithRetryAsync(request, cancellationToken).ConfigureAwait(false);
        string body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);

        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException(
                $"Could not generate a JIT config ({(int)response.StatusCode}). GitHub said:{Environment.NewLine}{body}");
        }

        JsonNode node = JsonNode.Parse(body)
            ?? throw new InvalidOperationException("The JIT config response was not JSON.");
        string blob = node["encoded_jit_config"]?.GetValue<string>()
            ?? throw new InvalidOperationException("The response contained no encoded_jit_config.");

        _logBus.RegisterSecret(blob);
        // The LENGTH only. The value must never reach a log.
        _logBus.Info("github", $"minted a JIT config for {runnerName} ({blob.Length} bytes)");

        return new JitConfig(runnerName, blob);
    }

    /// <summary>Verifies the App key by asking GitHub who we are.</summary>
    public async Task<string> TestAppCredentialsAsync(
        ForgeConfig config, CancellationToken cancellationToken = default)
    {
        string privateKey = _secretStore.Read("githubAppPrivateKey")
            ?? throw new InvalidOperationException("No GitHub App private key is stored.");

        string jwt = CreateAppJwt(config.GitHub.AppId, privateKey);

        using var request = new HttpRequestMessage(HttpMethod.Get, $"{ApiBaseUrl}/app");
        ApplyHeaders(request, "Bearer", jwt);

        using HttpResponseMessage response =
            await SendWithRetryAsync(request, cancellationToken).ConfigureAwait(false);
        string body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);

        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException(
                $"GET /app failed ({(int)response.StatusCode}). GitHub said:{Environment.NewLine}{body}");
        }

        JsonNode? node = JsonNode.Parse(body);
        return node?["slug"]?.GetValue<string>() ?? "(the app has no slug)";
    }

    /// <summary>Checks repository access, reporting 200/404/403 distinctly.</summary>
    public async Task<string> VerifyRepositoryAccessAsync(
        ForgeConfig config, string repo, CancellationToken cancellationToken = default)
    {
        string token = await GetInstallationTokenAsync(config, cancellationToken).ConfigureAwait(false);

        using var request = new HttpRequestMessage(
            HttpMethod.Get, $"{ApiBaseUrl}/repos/{config.GitHub.Owner}/{repo}");
        ApplyHeaders(request, "Bearer", token);

        using HttpResponseMessage response =
            await SendWithRetryAsync(request, cancellationToken).ConfigureAwait(false);

        return (int)response.StatusCode switch
        {
            200 => "OK — the installation can see this repository.",
            404 => "404 — either the repository does not exist, or the App is not installed on it. "
                   + "GitHub returns 404 rather than 403 for repositories you cannot see at all.",
            403 => "403 — the App is installed but lacks permission. It needs Administration: read & write "
                   + "to register self-hosted runners.",
            _ => $"{(int)response.StatusCode} {response.ReasonPhrase}",
        };
    }

    private static void ApplyHeaders(HttpRequestMessage request, string scheme, string token)
    {
        request.Headers.Authorization = new AuthenticationHeaderValue(scheme, token);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
        request.Headers.Add("X-GitHub-Api-Version", "2022-11-28");
        request.Headers.UserAgent.ParseAdd("RunnerForge/1.0");
    }

    /// <summary>
    /// Retries 5xx and secondary-rate-limit 403s with exponential backoff and
    /// honours Retry-After. Other statuses are returned as-is for the caller to
    /// surface verbatim.
    /// </summary>
    private async Task<HttpResponseMessage> SendWithRetryAsync(
        HttpRequestMessage request, CancellationToken cancellationToken)
    {
        const int maxAttempts = 4;

        for (int attempt = 1; ; attempt++)
        {
            using HttpRequestMessage attemptRequest = await CloneAsync(request).ConfigureAwait(false);
            HttpResponseMessage response =
                await _httpClient.SendAsync(attemptRequest, cancellationToken).ConfigureAwait(false);

            bool retryable = (int)response.StatusCode >= 500
                             || ((int)response.StatusCode == 403
                                 && response.Headers.RetryAfter is not null);

            if (!retryable || attempt >= maxAttempts) return response;

            TimeSpan delay = response.Headers.RetryAfter?.Delta
                             ?? TimeSpan.FromSeconds(Math.Pow(2, attempt));

            _logBus.Warning("github",
                $"{(int)response.StatusCode} from {request.RequestUri?.AbsolutePath}; "
                + $"retrying in {delay.TotalSeconds:0}s (attempt {attempt}/{maxAttempts})");

            response.Dispose();
            await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
        }
    }

    private static async Task<HttpRequestMessage> CloneAsync(HttpRequestMessage request)
    {
        var clone = new HttpRequestMessage(request.Method, request.RequestUri);

        if (request.Content is not null)
        {
            string body = await request.Content.ReadAsStringAsync().ConfigureAwait(false);
            clone.Content = new StringContent(body, Encoding.UTF8, "application/json");
        }

        foreach ((string key, IEnumerable<string> values) in request.Headers)
        {
            clone.Headers.TryAddWithoutValidation(key, values);
        }

        return clone;
    }
}

/// <summary>
/// A single-use runner registration. The blob never touches disk and never
/// appears in an argument list.
/// </summary>
public sealed record JitConfig(string RunnerName, string EncodedConfig);
